src/main/java/picard/sam/SamToFastq.java

```diff
@@ -168,15 +182,34 @@
             "value is null then all bases left after trimming will be written.", optional = true)
     public Integer READ2_MAX_BASES_TO_WRITE;
 
-    @Argument(shortName="Q", doc="End-trim reads using the phred/bwa quality trimming algorithm and this quality.", optional=true)
+    @Argument(shortName="Q", doc="End-trim reads using the phred/bwa quality trimming algorithm and this quality.", optional = true)
     public Integer QUALITY;
 
     @Argument(doc = "If true, include non-primary alignments in the output.  Support of non-primary alignments in SamToFastq " +
             "is not comprehensive, so there may be exceptions if this is set to true and there are paired reads with non-primary alignments.")
     public boolean INCLUDE_NON_PRIMARY_ALIGNMENTS = false;
 
+    @Argument(shortName = "STG", doc = "List of comma separated tag values to extract from Input SAM/BAM to be used as read sequence", optional = true)
+    public List<String> SEQUENCE_TAG_GROUP;
+
+    @Argument(shortName = "QTG", doc = "List of comma separated tag values to extract from Input SAM/BAM to be used as read qualities", optional = true)
+    public List<String> QUALITY_TAG_GROUP;
+
+    @Argument(shortName = "SEP", doc = "List of sequences to put in between each comma separated list of sequence tags (QFT)", optional = true)
+    public List<String> TAG_GROUP_SEPERATOR;
+
+    @Argument(shortName = "GZOPTG", doc = "Compress output FASTQ files per Tag grouping using gzip and append a .gz extension to the file names.")
+    public Boolean COMPRESS_OUTPUTS_PER_TAG_GROUP = false;
+
     private final Log log = Log.getInstance(SamToFastq.class);
 
+    private final static String TAG_SPLIT_DEFAULT_SEP = "";
+    private final static String TAG_SPLIT_DEFAULT_QUAL = "~";
+
+    private ArrayList<String[]> SPLIT_SEQUENCE_TAGS;
+    private ArrayList<String[]> SPLIT_QUALITY_TAGS;
+    private ArrayList<String> SPLIT_SEPARATOR_TAGS;

```

Comment:
This is just my opinion, but I feel like this tool is dying a death by a thousand cuts.  There are just so many options now, that I am have to read the code to know what it really does.  For example how does the newer option `OUTPUT_PER_RG` and this play together? 

How hard would it be to re-factor this tool so that you can implement another `SamToFastqWithTags` that sub-classes `SamToFastq`?  I worry we'll never detangle this code.

Reply:
As far as `OUTPUT_PER_RG` it would actually just also split the tag fastqs by readgroup as well.  The one weird case is wanting paired fastqs with tag fastqs because then they all wouldnt have the same number of lines.